documentation/projects/proposals/infrastructure_improvements/20240105_implementation_plan_move_remaining_services_to_openverse_org.md

```diff
@@ -0,0 +1,738 @@
+to notify registered applications of the change, including a management command
+to email them and a Make post to draft and publish.
+
+Potentially the most complexity aspect of this is actually the Cloudflare
+redirect rule to send API traffic from the `openverse.engineering` domains to
+`openverse.org`. I am 95% confident that we can do this with a single dynamic
+redirect rule, which is available in the free tier.
+
+The rule should match requests using something like the following expression:
+
+```
+http.host matches "api(-production)?\.openverse.engineering"
+```
+
+As mentioned in the step-by-step section, we will only redirect production
+domains. This reduces the complexity of the match expression and the dynamic
+target expression below. It also helps to clarify the fact that staging is
+volatile and has no guarantees for anyone using it (which is mostly us).
+
+The expression should also temporarily look for a special header's presence,
+acting as a feature flag, so that we don't immediately start to redirect normal
+users and can test the redirect:
+
+```
+any(http.request.headers.names[*] == "x-openverse-redirect") and http.host matches "api(-production)?\.openverse.engineering"
+```
+
+Then, because we are only redirecting production, the match expression is
+thankfully trivial and does not require `regex_replace` or other string
+manipulation:
+
+```
+concat("https://api.openverse.org", http.request.uri)
+```
+
+We cannot modify query parameters in the redirect, so there's nothing we can do
+to re-write the `url` parameter to the oembed endpoint. Luckily, that endpoint's
+request serializer ignores the hostname of the `url` parameter altogether, so it
+won't cause a backwards compatibility issue.
+
+As mentioned in the step-by-step section, we will configure this rule in
+[the `cloudflare` root module](#moving-cloudflare-access-and-the-new-cloudflare-root-module).
+
+### Finalizing the move
+
+The finalization steps handle cleaning up documentation and other non-critical
+references, as well as announcing the change on Make. Old links to production
+API will continue to work as redirects, so there's no need to update them right
+away.
+
+## Dependencies
+
+### Feature flags
+
+<!-- List feature flags/environment variables that will be utilised in the development of this plan. -->
+
+We will utilise the following temporary Terraform variables for certain stages
+of the work (all covered in detail above):
+
+- `configure_domains`
+- `create_security_group`
+
+These will be removed by the time their relevant steps are complete.
+
+### Infrastructure
+
+<!-- Describe any infrastructure that will need to be provisioned or modified. In particular, identify associated potential cost changes. -->
+
+This entire project is fundamentally an infrastructure one, with small changes
+to the API and frontend code to support it.
+
+As covered above, the effect of this is a savings of 200 USD per month, once we
```

Comment:
Small accuracy correction 😄 
```suggestion
As covered above, the effect of this is a savings of 250 USD per month, once we
```

Reply:
Really? I assumed we were billing annually. Are we billing annually for our `openverse.org` zone? If not, should we switch? Unless we have concrete plans to switch away from Cloudflare entirely within the year, it is a savings of 600 USD over the year to bill annually.